internal/services/topic.go

```diff
@@ -0,0 +1,166 @@
+		// canary topic already exists, check replicas assignments
+		log.Printf("The canary topic %s already exists\n", metadata[0].Name)
+
+		// if we scale up then scale down and then scale up again, the preferred leader are not elected immediately
+		// we should check current assignments, leaders and maybe forcing a leader election (not supported by Sarama right now)
+		// TODO
+
+		refresh = len(brokers) != len(metadata[0].Partitions)
+		if assignments, err = ts.alterTopic(len(metadata[0].Partitions), len(brokers)); err != nil {
+			log.Printf("Error altering topic %s: %v", metadata[0].Name, err)
+			return len(brokers), assignments, refresh, err
+		}
+		ts.checkTopic(len(brokers), metadata[0])
+	}
+	return len(brokers), assignments, refresh, err
+}
+
+func (ts *topicService) Close() {
+	log.Printf("Closing topic service")
+
+	err := ts.admin.Close()
+	if err != nil {
+		log.Printf("Error closing the Sarama cluster admin: %v", err)
+		os.Exit(1)
+	}
+	log.Printf("Topic service closed")
+}
+
+func (ts *topicService) createTopic(brokersNumber int) (map[int32][]int32, error) {
+	assignments, minISR := ts.assignments(0, brokersNumber)
+
+	v := strconv.Itoa(int(minISR))
+	topicConfig := map[string]*string{
+		"min.insync.replicas": &v,
+	}
+
+	topicDetail := sarama.TopicDetail{
+		NumPartitions:     -1,
+		ReplicationFactor: -1,
+		ReplicaAssignment: assignments,
+		ConfigEntries:     topicConfig,
+	}
+	err := ts.admin.CreateTopic(ts.canaryConfig.Topic, &topicDetail, false)
+	return assignments, err
+}
+
+func (ts *topicService) alterTopic(currentPartitions int, brokersNumber int) (map[int32][]int32, error) {
+	assignments, _ := ts.assignments(currentPartitions, brokersNumber)
+
+	ass := make([][]int32, len(assignments))
+	for i := 0; i < len(ass); i++ {
+		ass[i] = make([]int32, len(assignments[int32(i)]))
+		copy(ass[i], assignments[int32(i)])
+	}
+	log.Printf("%v", ass)
+
+	var err error
+	// less partitions than brokers (scale up)
+	if currentPartitions < brokersNumber {
+		// passing the assigments just for the partitions that needs to be created
+		err = ts.admin.CreatePartitions(ts.canaryConfig.Topic, int32(brokersNumber), ass[currentPartitions:], false)
+	} else {
+		// more or equals partitions than brokers, just need reassignment
+		err = ts.admin.AlterPartitionReassignments(ts.canaryConfig.Topic, ass)
+	}
+	return assignments, err
+}
+
+func (ts *topicService) checkTopic(brokersNumber int, metadata *sarama.TopicMetadata) {
+	electLeader := false
+	if len(metadata.Partitions) == brokersNumber {
+		for _, p := range metadata.Partitions {
+			if p.ID != p.Leader {
+				electLeader = true
+			}
+		}
+	}
+	log.Printf("Elect leader = %t\n", electLeader)
+}
+
+func (ts *topicService) assignments(currentPartitions int, brokersNumber int) (map[int32][]int32, int) {
+	partitions := math.Max(float64(currentPartitions), float64(brokersNumber))
+	replicationFactor := math.Min(float64(brokersNumber), 3)
+	minISR := math.Max(1, replicationFactor-1)
```

Comment:
It looks like we're using floating point just so we can use `math.Max`. It would be cleaner imho you just write your own `max` function and stick with integer arithmetic.

Reply:
Yeah, that's one of the problems with Go where you don't have an "overload" thing but you can just play around with functions with different names. I will write my own Max, anyway, we are going to use just integers.